drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
-                } else {
-                    let status =
-                        self.transceivers.get_i2c_status(port_loc).map_err(
-                            |_e| Error::ReadFailed(HwError::ReadBufFailed),
-                        )?;
-
-                    if status & Reg::QSFP::PORT0_STATUS::BUSY == 0 {
-                        if status & Reg::QSFP::PORT0_STATUS::ERROR != 0 {
-                            failure_mask.set(port_loc.port.get())?;
-                        }
-                        break;
+                if self
+                    .transceivers
+                    .get_i2c_status_and_read_buffer(
+                        port_loc,
+                        &mut buf[0..(buf_len + 1)],
+                    )
+                    .is_err()
+                {
+                    error.set(port);
+                    break;
+                };
+
+                let status = buf[0];
+
+                // Use QSFP::PORT0 for constants, since they're all identical
+                if status & Reg::QSFP::PORT0_STATUS::BUSY == 0 {
+                    // Check error mask
+                    if status & Reg::QSFP::PORT0_STATUS::ERROR != 0 {
+                        // Record which port the error ocurred at so we can
+                        // give the host a more meaningful error.
+                        failure.set(port);
+                    } else {
+                        // Add data to payload
+                        success.set(port);
+                        let end_idx = idx + buf_len;
+                        out[idx..end_idx].copy_from_slice(&buf[1..][..buf_len]);
+                        idx = end_idx;
                     }
+                    break;
                 }
 
                 userlib::hl::sleep_for(1);
             }
         }
-        if failure_mask.selected_transceiver_count() != 0 {
-            return Err(Error::ReadFailed(HwError::I2cError(failure_mask)));
-        }
-        Ok(())
+        let final_result = ModuleResult::new(success, failure, error).unwrap();
+        result.chain(final_result)
     }
 
+    // The `LogicalPortMask` indicates which of the requested ports the
+    // `HwError` applies to.
     fn write(
         &mut self,
         mem: MemoryWrite,
-        modules: ModuleId,
+        modules: LogicalPortMask,
         data: &[u8],
-    ) -> Result<(), Error> {
-        let mask = get_mask(modules)?;
-
-        self.select_page(*mem.page(), mask)
-            .map_err(Error::WriteFailed)?;
+    ) -> ModuleResult {
+        let mut result = self.select_page(*mem.page(), modules);
 
         // Copy data into the FPGA write buffer
-        self.transceivers
-            .set_i2c_write_buffer(&data[..mem.len() as usize])
-            .map_err(|_e| Error::WriteFailed(HwError::WriteBufFailed))?;
+        result = result.chain(
+            self.transceivers
+                .set_i2c_write_buffer(&data[..mem.len() as usize]),
+        );
```

Comment:
If I'm understanding `set_i2c_write_buffer()` correctly, I don't think we want to call `chain()` like this. The `set_i2c_write_buffer()` function appears to push data into the write buffers for _all_ transceivers, independent of the value of `modules`. So, unless one of them fails that operation, we'll return and then overwrite `result.success` with _all_ modules.

I can see this happening I think, by looking at either the host-side log messages, or the `ringbuf` entries. For example:

```
ben@niles ~ $ echo '02' | ./xcvradm -i axf4 -t 0 -l trace write-lower --sff 93 --input-kind hex
Mar 31 04:23:43.345 DEBG bound UDP socket, local_addr: Ok([::]:41304), interface: axf4
Mar 31 04:23:43.345 DEBG joined IPv6 multicast group, multicast_addr: ff02::1de:2
Mar 31 04:23:43.345 DEBG spawned IO task
Mar 31 04:23:43.345 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 0, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x1), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: false, is_tx_task_set: false } }) } }, task: io
Mar 31 04:23:43.345 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x1), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, peer: [ff02::1de:2%2]:11112, task: io
Mar 31 04:23:43.350 TRCE packet received, peer: [fe80::aa40:25ff:fe05:ff00%2]:11112, n_bytes: 34, task: io
Mar 31 04:23:43.350 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Read { modules: ModuleId(0x1), failed_modules: ModuleId(0x0), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:ff00%2]:11112, task: io
Mar 31 04:23:43.351 TRCE received trailing data, n_bytes: 1, data: Some([17]), task: io
Mar 31 04:23:43.351 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 1, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Write { modules: ModuleId(0x1), write: MemoryWrite { page: Sff8636(Lower), offset: 93, len: 1 } }) }, data: Some([2]) }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Mar 31 04:23:43.351 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Write { modules: ModuleId(0x1), write: MemoryWrite { page: Sff8636(Lower), offset: 93, len: 1 } }) }, peer: [ff02::1de:2%2]:11112, task: io
Mar 31 04:23:43.361 TRCE packet received, peer: [fe80::aa40:25ff:fe05:ff00%2]:11112, n_bytes: 33, task: io
Mar 31 04:23:43.362 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Write { modules: ModuleId(0xffffffff), failed_modules: ModuleId(0x0), write: MemoryWrite { page: Sff8636(Lower), offset: 93, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:ff00%2]:11112, task: io
Mar 31 04:23:43.362 TRCE received trailing data, n_bytes: 0, data: Some([]), task: io
```

In that second-to-last log line, we print the `SpResponse::Write` we got back, which does indeed show all 1's for the written modules. The `ringbuf` shows the same:

```
ben@niles ~ $ pfexec humility --target sidecar-b-aaron ringbuf transceivers
humility: attached to 0483:3754:002600184D4B500E20373831 via ST-Link V3
humility: ring buffer drv_transceivers_server::__RINGBUF in transceivers:
 NDX LINE      GEN    COUNT PAYLOAD
   6  238        6        1 GotInterface(0x1, Sff8636)
   7  252        6        1 TemperatureReadError(0x2, ImplError(0x4))
   8  238        6        1 GotInterface(0x3, Sff8636)
   9  238        6        1 GotInterface(0x0, Sff8636)
  10  238        6        1 GotInterface(0x2, Sff8636)
  11  238        6        1 GotInterface(0x4, Sff8636)
  12  238        6        1 GotInterface(0x6, Sff8636)
  13  238        6        1 GotInterface(0x8, Sff8636)
  14  238        6        1 GotInterface(0xa, Sff8636)
  15  238        6        1 GotInterface(0xc, Sff8636)
   0  238        7        1 GotInterface(0xe, Sff8636)
   1  238        7        1 GotInterface(0x10, Sff8636)
   2  238        7        1 GotInterface(0x12, Sff8636)
   3  238        7        1 GotInterface(0x14, Sff8636)
   4  238        7        1 GotInterface(0x18, Sff8636)
   5  238        7        1 GotInterface(0x1a, Sff8636)
humility: ring buffer drv_transceivers_server::udp::__RINGBUF in transceivers:
 NDX LINE      GEN    COUNT PAYLOAD
   3  272       95        1 ResponseSize(ResponseSize { header_length: 0xa, message_length: 0x17, data_length: 0x3 })
   4  319       95        1 Read(ModuleId(0x7), MemoryRead { page: Sff8636(Lower), offset: 0x5d, len: 0x1 })
   5  332       95        1 OperationResult(ModuleResult { success: LogicalPortMask(0x7), failure: LogicalPortMask(0x0), error: LogicalPortMask(0x0) })
   6  272       95        1 ResponseSize(ResponseSize { header_length: 0xa, message_length: 0x17, data_length: 0x3 })
   7  319       95        1 Read(ModuleId(0x1), MemoryRead { page: Sff8636(Lower), offset: 0x0, len: 0x1 })
   8  332       95        1 OperationResult(ModuleResult { success: LogicalPortMask(0x1), failure: LogicalPortMask(0x0), error: LogicalPortMask(0x0) })
   9  272       95        1 ResponseSize(ResponseSize { header_length: 0xa, message_length: 0x17, data_length: 0x1 })
  10  353       95        1 Write(ModuleId(0x1), MemoryWrite { page: Sff8636(Lower), offset: 0x5d, len: 0x1 })
  11  365       95        1 OperationResult(ModuleResult { success: LogicalPortMask(0xffffffff), failure: LogicalPortMask(0x0), error: LogicalPortMask(0x0) })
  12  272       95        1 ResponseSize(ResponseSize { header_length: 0xa, message_length: 0x17, data_length: 0x0 })
  13  319       95        1 Read(ModuleId(0x1), MemoryRead { page: Sff8636(Lower), offset: 0x0, len: 0x1 })
  14  332       95        1 OperationResult(ModuleResult { success: LogicalPortMask(0x1), failure: LogicalPortMask(0x0), error: LogicalPortMask(0x0) })
  15  272       95        1 ResponseSize(ResponseSize { header_length: 0xa, message_length: 0x17, data_length: 0x1 })
   0  353       96        1 Write(ModuleId(0x1), MemoryWrite { page: Sff8636(Lower), offset: 0x5d, len: 0x1 })
   1  365       96        1 OperationResult(ModuleResult { success: LogicalPortMask(0xffffffff), failure: LogicalPortMask(0x0), error: LogicalPortMask(0x0) })
   2  272       96        1 ResponseSize(ResponseSize { header_length: 0xa, message_length: 0x17, data_length: 0x0 })
ben@niles ~ $
```